packages/cli/src/modules/mcp/tools/execute-workflow.tool.ts

```diff
@@ -0,0 +1,275 @@
+				})
+				.optional()
+				.describe('Input data for webhook-based workflows'),
+		})
+		.optional()
+		.describe('Inputs to provide to the workflow'),
+});
+
+type ExecuteWorkflowOutput = {
+	success: boolean;
+	executionId: string | null;
+	result?: IRunExecutionData['resultData'];
+	error?: unknown;
+};
+
+const outputSchema = {
+	success: z.boolean(),
+	executionId: z.string().nullable().optional(),
+	result: z.any().optional().describe('Workflow execution result data'),
+	error: z.unknown().optional(),
+} satisfies z.ZodRawShape;
+
+export const createExecuteWorkflowTool = (
+	user: User,
+	workflowFinderService: WorkflowFinderService,
+	activeExecutions: ActiveExecutions,
+	workflowRunner: WorkflowRunner,
+	telemetry: Telemetry,
+): ToolDefinition<typeof inputSchema.shape> => ({
+	name: 'execute_workflow',
+	config: {
+		description: 'Execute a workflow by id',
+		inputSchema: inputSchema.shape,
+		outputSchema,
+	},
+	handler: async ({ workflowId, inputs }) => {
+		const telemetryPayload: UserCalledMCPToolEventPayload = {
+			user_id: user.id,
+			tool_name: 'execute_workflow',
+			parameters: { workflowId, inputs },
+		};
+		try {
+			const output = await executeWorkflow(
+				user,
+				workflowFinderService,
+				activeExecutions,
+				workflowRunner,
+				workflowId,
+				inputs,
+			);
+
+			telemetryPayload.results = {
+				success: output.success,
+				data: {
+					executionId: output.executionId,
+				},
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: JSON.stringify(output) }],
+				structuredContent: output,
+			};
+		} catch (error) {
+			const payload: ExecuteWorkflowOutput = {
+				success: false,
+				executionId: null,
+				error,
+			};
+
+			telemetryPayload.results = {
+				success: false,
+				error: error instanceof Error ? error.message : String(error),
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: JSON.stringify(payload) }],
+				structuredContent: payload,
+			};
+		}
+	},
+});
+
+/**
+ * Executes a workflow for the given user with provided inputs.
+ * In order to "synchronously" execute the workflow,
+ * it is mapping mcp tool inputs to trigger node pin data and starting execution from there.
+ * LIMITATION: Does not properly support workflows with multiple triggers.
+ */
+export const executeWorkflow = async (
+	user: User,
+	workflowFinderService: WorkflowFinderService,
+	activeExecutions: ActiveExecutions,
+	workflowRunner: WorkflowRunner,
+	workflowId: string,
+	inputs?: z.infer<typeof inputSchema>['inputs'],
+): Promise<ExecuteWorkflowOutput> => {
+	const workflow = await workflowFinderService.findWorkflowForUser(workflowId, user, [
+		'workflow:read',
+	]);
+
+	if (!workflow || workflow.isArchived || !workflow.settings?.availableInMCP) {
+		throw new UserError('Workflow not found');
+	}
```

Comment:
Should we be more granular about this, as in, should we tell the MCP client exactly what happened so the client can relay to the user exactly what happened? I'd be confused if the workflow exists but it's failing because another user in the interim set it to not available in MCP.

Reply:
Yeah, probably it makes sense to differentiate between `!availableInMCP` and others.